ultravox/model/ultravox_model.py

```diff
@@ -186,17 +190,29 @@ def forward(
                 len(audio_token_start_idx) == len(audio_token_len) == len(audio_values)
             ), "audio_token_start_idx, audio_token_len, and audio_values must have the same batch size."
 
-            # B x A/3200 x D
-            audio_tower_output = self.audio_tower.forward(
-                audio_values
-            ).last_hidden_state
-            audio_tower_output = audio_tower_output.to(inputs_embeds.dtype)
+            # Check if the audio_values T dimension is greater than whisper encoder's context window.
+            if audio_values.shape[2] > self.audio_tower_context_length:
+                audio_values_chunks = torch.split(
+                    audio_values, self.audio_tower_context_length, dim=2
+                )
```

Comment:
Huh, this message was removed somehow. You must've been so confused!

Okay so one approach I think we could take to handle both problems is:
- handle breaking up the audio in processor instead (it might already have some methods for doing so, not a big deal if not either)
- now we can't assume audio_embeds and input_embeds are matched in their index. The collator should make sure to correctly index all the audio chunks and provide the index to the forward function
- the forward function doesn't need to change much, just that it takes the `audio_batch_index` as an input and uses that to know where to put the audio_embeds instead of assuming the indices match

Reply:
I agree this is not the right place to handle this.

The length restriction of the whisper encoder should be handled in the forward method of the whisper encoder.

Reply:
I don't think that's how Whisper works (e.g. see how [ASR pipeline](https://github.com/huggingface/transformers/blob/a05ce550bf1cda8ae3bffcc62607597770460a42/src/transformers/pipelines/automatic_speech_recognition.py#L350) handles this). It relies on processor to create the right-sized chunks.

From what I've seen in `transformers`, usually things like truncation, padding, and splitting are handled by processor to allow for better configurability and to not make the model too complicated.

Reply:
Another reason why doing this in forward function is bad (be it Ultravox.forward or Whisper.forward), is that padding happens first and then things get chunked.

That means if you have one sample that is too big in a batch, all of your samples will have a lot of chunks.

Reply:
I think you are right, it should be first split in the processor. Let's discuss offline.